### PR TITLE
Add SubstsBuilder

### PR DESCRIPTION
**Pull request by flodiebold**
+ further refactoring.

### CODE REVIEW

**Comment by flodiebold** [comment]
It seems Travis doesn't use 1.38 yet...

**Comment by matklad** [comment]
Ah, I think I need to bust the caches for this...

File: crates/ra_hir/src/code_model.rs [hunk 1]
```diff
@@ -339,10 +339,14 @@ pub struct Struct {
...
 
 impl Struct {
-    pub fn module(self, db: &impl HirDatabase) -> Module {
+    pub fn module(self, db: &impl DefDatabase) -> Module {
         self.id.module(db)
     }
 
+    pub fn krate(self, db: &impl DefDatabase) -> Option<Crate> {
+        self.module(db).krate(db)
+    }
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]: Heh, we really should add some traits around these

Comment (matklad) [reply crates/ra_hir/src/code_model.rs]: OTOH, I am wondering if the absolutely huge hir compile times are due to our use of traints and Froms?

Comment (matklad) [reply crates/ra_hir/src/code_model.rs]: See https://github.com/EmbarkStudios/rust-ecosystem/issues/42#issuecomment-533822802

File: crates/ra_hir/src/ty.rs [hunk 2]
```diff
@@ -111,6 +111,80 @@ pub enum TypeCtor {
...
+            | TypeCtor::Char
+            | TypeCtor::Int(_)
+            | TypeCtor::Float(_)
+            | TypeCtor::Str
+            | TypeCtor::Never => 0,
+            TypeCtor::Slice
+            | TypeCtor::Array
+            | TypeCtor::RawPtr(_)
+            | TypeCtor::Ref(_)
+            | TypeCtor::Closure { .. } => 1,
```
Comment (matklad) [inline crates/ra_hir/src/ty.rs]: That's the `Self` type of closure? Maybe a good place to drop a comment

Comment (flodiebold) [reply crates/ra_hir/src/ty.rs]: It's the signature

File: crates/ra_hir/src/code_model.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/code_model.rs]: Yeah, I was thinking about doing that as well.

**Pull request merged**